### ISSUE TITLE
Cyanogenmod + microG

### ISSUE THREAD

**Issue by okias** [microg/android_packages_apps_GmsCore#63]
Hey guys, I decided to start discussion about integrating with Cyanogenmod.

I think it's important to negotiate some solution, because CM build could be big oportunity to make uG useful. Custom patching of course making it harder.

I just asked question http://review.cyanogenmod.org/#/c/106162/ hope it's correct place to discussion.

I'm sure, there can be found some solution for this challenge.

**Comment by bonanza123** [comment]
Just my opinion: It's a waste of effort as the 'original' CM and its idea is dead, at least since they decided to become a company. OmniRom is the alternative, many former CM developers switched to it.

**Comment by okias** [comment]
@bonanza123 Well, still I have feeling they have better ecosystem than AOSP :) and for end-user, CM is still better option

**Comment by binaricode** [comment]
Maybe they would accept it if the patch was written to only work if an option had been enabled in the hidden "Developer options" settings. That would prevent your average user from accidentally compromising security.

**Comment by Catfriend1** [comment]
hi,i am also trying to use UG with cm Rom. it works but the apps depending on gms do not recognize that gms are installed. tried Android patch but no solution found yet. its partly Working on cm12.1 with Samsung i9100. wrote some problems here;: https://github.com/microg/android_packages_apps_UnifiedNlp/issues/39

**Comment by Catfriend1** [comment]
Ok, got it working by installing XPosed Framework and FakeGApps. The App Push Notifications Tester from Play Store now recognizes MicroG as Google GMS and GCM is working fine.

**Comment by okias** [comment]
Recreated better request including changes accepted in OmniROM:

http://review.cyanogenmod.org/#/c/125635/
http://review.cyanogenmod.org/#/c/125636/

**Comment by okias** [comment]
(modified code to fit, hope it'll compile with CM-12.1)

**Comment by mar-v-in** [comment]
As @okias obviously noted there is an unreasonable strong opinion against the feature backed into CM. It should work fine with XPosed + FakeGApps, however XPosed is huge and can cause serious problems, so there should always be a way to do things without XPosed if possible.

@moosd wrote a nice and small python script [1] to apply the patch after building and flashing the rom. It currently requires a desktop connected to the phone for patching, but ideally we are able to put this into a root app. As CM ships with root support, this would allow users to patch their cm installation to include signature faking support with less pain. Ideally we would also detect system updates and re-apply the patch automatically.

[1] https://github.com/moosd/Needle

**Comment by Catfriend1** [comment]
+1 for the App :-)
i like this idea. having Seen what Google stores in partly encrypted gms database i am with you. -1 for cm sticking with Google. for me, they would not make their hands dirty when a normal user can decide against the force of cloud. Google bound out Bands very tight too its services. e.g. play store licenses and moving more ad and tracking services to their gms even updating gms silently and encrypted through firewall or mobile connection.

**Comment by pedro-araujo** [comment]
@mar-v-in About the script from @moosd , is it necessary to fake the app signature for each individual app like he describes later in the README or patching the framework is more than enough?

**Comment by moosd** [comment]
@pedro-araujo AFAIK play services doesn't check the signature of apps that connect to it (it has no idea what these signatures are meant to be, after all) - but the apps connecting to it check the signature of play-services (since it's only ever going to be signed by Google).

The bit later in the readme is just for completeness, in case you wish to fake the signature of any arbitrary app (e.g. if it verifies its own signature or something). It's already done for the microg play services, so patching the framework is enough for it to work.

**Comment by pedro-araujo** [comment]
@moosd thank you for the reply, I understand now. I was asking it because I tried to use the script with CM13 and I keep having the same invalid signature errors in the logcat and I was wondering if I had skip any step.

I guess it might be just incompatible. I will open a bug report on Needle's repository if that's alright with you.

**Comment by mar-v-in** [comment]
@moosd's Needle is now linked in README and CyanogenMod administration clearly stated that they will never allow any signature spoofing as part of their ROM.

There is still the possibility to create a special version of GmsCore for CyanogenMod that does not require signature spoofing but uses the fact that signatures are not properly verified on system apps. However, patching the system via Xposed or with Needle is a nice solution as well and has several advantages over that hacky solution.

As a solution for CyanogenMod we can later (whatever that means) provide an android root app to patch the system (the same way as Needle does, but ideally in a way that it can be applied automatically after update). I will leave this issue open for that reason

**Comment by aphirst** [comment]
@mar-v-in Is there anything yet in the works, either from this project or from others, for the signature spoofing that would allow microG **and only microG** to spoof being GApps (and perhaps, **only GApps**), rather than allowing spoofing across-the-board?

**Comment by mar-v-in** [comment]
The Xposed module does only spoof the currently installed Play Store and Play Services to be Google one without allowing other apps to spoof. The signature spoofing used in OmniROM 6.0 requires a confirmation for each app (based on Android 6 dynamic permissions). This and additional restrictions (by package name, system apps only, ...) were suggested and I would develop them in case CM would agree to include them (in my opinion it's waste of time, as by-app confirmation already provides very high security). 

However CM core team denied including any patch that will allow to spoof signature, independent from how strong it is secured. The proposal to secure the feature the same way as root permissions are protected (a tick in developer options and per app confirmation) was denied still arguing with security. Hinting them on the fact that enforcing a higher security level than root permissions is stupid as obviously you can patch the system once you got root permission and thus have that feature was countered with a sarcastic "we wan't to remove root permissions anyway, it's only there for legacy reasons".

Summary: I am willing to develop a modified version of the patch that has a better/other way to restrict the feature to ensure high security if that would be interesting for anyone building a publicly accessible custom ROM.

Note that I think it's a bad idea to restrict this feature to certain packages or authorities, as there are not a lot ways to misuse it, it might be required for more packages in the future and authorities is just another word for vendor locking, one of the things this project is actually working against: Everyone should be able to build a custom version of GmsCore and use it

**Comment by aphirst** [comment]
@mar-v-in Alright - glad to hear your thoughts and to know that is is something that's in developer awareness!

**Comment by motatuc** [comment]
Hello

I applied the patches for CM12.1 that are provided here: http://review.cyanogenmod.org/#/c/125635/ and here: http://review.cyanogenmod.org/#/c/125636/. These two work like they should after signature spoofing gets enabled in the developer settings.

Yesterday I tried to applied the patches for CM13 that can be found here: http://review.cyanogenmod.org/#/c/125637/

Do I have to enable them somewhere as well? Because they do not seem to work.
I installed microG GmsCore but the spoofing does not work. 

Is this patch/commit tested for CM13 or did some things change?
